pkg/ruler/ruler_pagination.go

```diff
@@ -0,0 +1,42 @@
+package ruler
+
+import (
+	"crypto/sha1"
+	"encoding/hex"
+)
+
+type GroupStateDescs []*GroupStateDesc
+
+func (gi GroupStateDescs) Swap(i, j int) { gi[i], gi[j] = gi[j], gi[i] }
+func (gi GroupStateDescs) Less(i, j int) bool {
+	return GetRuleGroupNextToken(gi[i].Group.Namespace, gi[i].Group.Name) < GetRuleGroupNextToken(gi[j].Group.Namespace, gi[j].Group.Name)
+}
+func (gi GroupStateDescs) Len() int { return len(gi) }
+
+func GetRuleGroupNextToken(namespace string, group string) string {
+	h := sha1.New()
+	h.Write([]byte(namespace + ";" + group))
+	return hex.EncodeToString(h.Sum(nil))
+}
+
+func TruncateGroups(groups []*GroupStateDesc, maxRuleGroups int) ([]*GroupStateDesc, string) {
```

Comment:
do these functions need to be public?

Reply:
I'm talking only about TruncateGroups and GetRuleGroupNextToken

Reply:
`TruncateGroups` does not need to be public. I will fix it. `GetRuleGroupNextToken` needs to be public since it is used in integration testing

Reply:
I would actually move all the content of this file to close where they are used, since they are only used in one place. I'm not particularly strong about this but I think it would be simpler to understand.

Reply:
Another suggestion, instead of truncateGroups we could name this function generatePage, createPage, makePage or similar to make it clear that this function is used to generate the page response with token etc. I think this function deserves some comments describing how the pagination works and that is supposed to receive a sorted list of groups. just my two cents.